compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs

```diff
@@ -2837,26 +2846,172 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
+                            if let Ok(_) = c.relate(trait_pred, predicate) {
+                                type_diffs = c.errors;
+                            }
+                        }
+                    }
+                    let point_at_chain = |expr: &hir::Expr<'_>| {
+                        let mut expr = expr;
+                        let mut prev_ty = self.resolve_vars_if_possible(
+                            typeck_results.expr_ty_adjusted_opt(expr).unwrap_or(tcx.ty_error()),
                         );
+                        let outer_ty = prev_ty;
+                        let mut assoc_seen = 0;
+                        while let hir::ExprKind::MethodCall(_path_segment, rcvr_expr, _args, span) =
+                            expr.kind
+                        {
+                            // Point at every method call in the chain with the resulting type.
+                            // vec![1, 2, 3].iter().map(mapper).sum<i32>()
+                            //               ^^^^^^ ^^^^^^^^^^^
+                            expr = rcvr_expr;
+
+                            let ocx = ObligationCtxt::new_in_snapshot(self.infcx);
+                            for diff in &type_diffs {
+                                let Sorts(expected_found) = diff else { continue; };
+                                let ty::Projection(proj) = expected_found.expected.kind() else { continue; };
+                                assoc_seen += 1;
+
+                                let origin = TypeVariableOrigin {
+                                    kind: TypeVariableOriginKind::TypeInference,
+                                    span,
+                                };
+                                let trait_def_id = proj.trait_def_id(self.tcx);
+                                let substs =
+                                    InternalSubsts::for_item(self.tcx, trait_def_id, |param, _| {
+                                        match param.kind {
+                                            ty::GenericParamDefKind::Type { .. } => {
+                                                if param.index == 0 {
+                                                    return prev_ty.into();
+                                                }
+                                            }
+                                            ty::GenericParamDefKind::Lifetime
+                                            | ty::GenericParamDefKind::Const { .. } => {}
+                                        }
+                                        self.var_for_def(span, param)
+                                    });
+                                let ty_var = self.infcx.next_ty_var(origin);
+                                let trait_ref = ty::Binder::dummy(ty::PredicateKind::Clause(
+                                    ty::Clause::Projection(ty::ProjectionPredicate {
+                                        projection_ty: ty::ProjectionTy { substs, ..*proj },
```

Comment:
nit: better to be explicit here

```suggestion
                                        projection_ty: ty::ProjectionTy { item_def_id: proj.item_def_id, substs },
```